docs/architecture/architecture.md

```diff
@@ -0,0 +1,378 @@
+
+1. `cc-runtime` connects to `cc-proxy` and sends it the `attach` command to let it know
+on which pod the container it is trying to to `delete` is running.
+2. `cc-runtime` sends an agent `DESTROYPOD` command to `destroy` the pod holding the
+container running on the guest. The command is sent to `cc-proxy` who forwards it to the right
+agent instance running in the appropriate guest.
+3. After deleting the last running pod, the `agent` will gracefully shut the virtual machine
+down.
+4. `cc-runtime` sends the `BYE` command to `cc-proxy`, to let it know that a given virtual
+machine is shut down. `cc-proxy` will then clean all its internal resources associated with this
+VM.
+
+#### Proxy
+
+`cc-proxy` is a daemon offering access to the VM [`agent`](https://github.com/clearcontainers/agent)
+to multiple `cc-shim` and `cc-runtime` clients.
+Only a single instance of `cc-proxy` per host is necessary as it can be used for several different VMs.
+Its main role is to:
+- Arbitrate access to the `agent` control channel between all the `cc-runtime` instances and the `cc-shim` ones.
+- Route the I/O streams between the various `cc-shim` instances and the `agent`.
+
+`cc-proxy` provides 2 client interfaces:
+
+- A UNIX, named socket for all `cc-runtime` instances on the host to send commands to `cc-proxy`.
+- One socket pair per `cc-shim` instance, to send stdin and receive stdout and stderr I/O streams. See the
+[cc-shim section](#shim)
+for more details about that interface.
+
+The protocol on the `cc-proxy` UNIX named socket supports the following commands:
+
+- `Hello`: This command is for `cc-runtime` to let `cc-proxy` know about a newly created VM that will
+hold containers. This command payload contains the `agent` control and I/O UNIX socket paths created
+and exported by QEMU, and `cc-proxy` will connect to both of them after receiving the `Hello` command.
+- `Bye`: This is the opposite of `Hello`, i.e. `cc-runtime` uses this command to let `cc-proxy` know
+that it can release all resources related to the VM described in the command payload.
+- `Attach`: `cc-runtime` uses that command as a VM multiplexer as it allows it to notify `cc-proxy` about
+which VM it wants to talk to. In other words, this commands allows `cc-runtime` to attach itself to a
+running VM.
+- `AllocateIO`: As the `agent` can potentially handle I/O streams from multiple container processes at the
+same time, it needs to be able to associate any given stream to a container process. This is done by the `agent`
+allocating a set of at most 2 so-called sequence numbers per container process. `cc-runtime` will send
+the `AllocateIO` command to `cc-proxy` to have it request the `agent` to allocate those sequence numbers.
+They will be passed as command line arguments to `cc-shim`, who will then use them to e.g. prepend its stdin
+stream packets with the right sequence number.
+- `Hyper`: This command is used by both `cc-runtime` and `cc-shim` to forward `agent` specific
```

Comment:
@jcvenegas - can you review this section?

Reply:
This is well out of date as well for 3.0.

Reply:
Updated to 3.0 proxy protocol in wiki